[otbn] Add modular multiplication with Barrett reduction

felixmiller opened the pull request:
This is a generic implementation without requirements on the form
of the modulus, only the range of the modulus is limited. For
special formed moduli (e.g. generilzed Mersenne primes in p384)
we might want to provide an optimized version.
The implementation mainly follows the description of the algorithm
in the "Handbook of Applied Cryptography" (Algorithm 14.42) and
re-uses its terminology. Some changes have been made to improve
register and multiplier efficiency.

This is related to lowRISC#3192.

Signed-off-by: Felix Miller <felix.miller@gi-de.com>

rswarbrick commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,254 @@
+  // neglected to allow using a 384x384 multiplier. For the MSb of x being set we temporary store u (or zero) here to
+  // be used in a later constant time correction of a multiplication with u. Note that this requires the MSb flag
+  // being carried over from the multiplication routine.
+  BN.SEL w24, w14, w31, m
+  BN.SEL w25, w15, w31, m
```
The flag group needs capitalising (`M`)

felixmiller replied:
done

rswarbrick commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,254 @@
+  // Conditionally add 2^385 to r if r is negative.
+  // r = [w22, w21] <= (r < 0)?r + 2^385 = [w22, w21] + 2^385 = [w22 + 2^129, w21]
+  BN.ADD w30, w30, w22
+  BN.SEL w22, w30, w22, c ;
```
Again, the flag group needs capitalising here and on the next few lines. Also, we don't support `;` as a comment character.

felixmiller replied:
done

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,254 @@
+ *   Constant time 384 bit modular multiplication based on Barrett reduction algorithm.
+ *
+ *   Provides a generic modular multiplication kernel for 384 bit wide numbers without requiring a special form of the
+ *   modulus.
```
Changed this, actually thought the delimiter in my IDE was set to 100 but was actually 120. Changed to 79 now.

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,254 @@
+  BN.XOR w18, w18, w18
+  BN.XOR w20, w20, w20
+
+  // [w17, w16] = a[255:0] x b[255:0] = w[8] * w[10]  (512=256x256)
```
done

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,254 @@
+  // [w17, w16] = a[255:0] x b[255:0] = w[8] * w[10]  (512=256x256)
+  BN.MULQACC.Z w8.0, w10.0, 0
+  BN.MULQACC w8.1, w10.0, 64
+  BN.MULQACC.SO w16.l, w8.0, w10.1, 64
```
done

felixmiller commented:
Thanks @rswarbrick for having a look at this. Did all the changes as requested, otbn-as is happy now

felixmiller commented:
On thing to note: The code consumes quite some registers (naturally more than the 256 bit variant). This can be probably tweaked to spare maybe two regs by changing the order of access. Created #3267 for this. This should, however have no effect on performance evaluation.

imphil commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+ * Differences:
+ *   - The computation of q2 ignores the MSbs of q1 and u to allow using
+ *     a 384x384 bit multiplication. This is compensated later by
+       individual (conditional) additions.
```
```suggestion
 *     individual (conditional) additions.
```

felixmiller replied:
done

imphil commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+ * most 2^385-1, it has to be ensured that m >= 2^383 + 1.
+ *
+ * This implementation mostly follows the description in the
+ * "Handbook of applied Cryptography" in Algorithm 14.42.
```
```suggestion
 * "Handbook of Applied Cryptography" in Algorithm 14.42.
```

felixmiller replied:
done

imphil reviewed (commented):
Two extreme nitpicky comments from my side, but looks fantastic! Leaving the final review to @rswarbrick.

rswarbrick commented:
Hi! Sorry for being a bit slow: I'll try to have a proper look at this before the end of the day.

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+  BN.ADD w17, w17, w19
+  BN.ADDC w18, w18, w31
+
+  /* [w18, w17] = b[255:0] x a[383:256] = w[10] * w[9]  (384=256x256) */
```
typo, should be (384=256x128), will fix this

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+  BN.MULQACC.SO w17.l, w8.2, w10.3, 64
+  BN.MULQACC.SO w17.u, w8.3, w10.3, 0
+
+  /* [w18, w19] = a[255:0] x b[383:256] = w[8] * w[11]  (384=256x256) */
```
typo, should be (384=256x128), will fix this

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+ *                               (without u[384]/MSb of u which is always 1
+ *                               for the allowed range but has to be set to 0
+ *                               here).
+ * @param[out] [559:512]: c, result, max. length 384 bit.
```
typo
```suggestion
 * @param[out] dmem[559:512]: c, result, max. length 384 bit.
```

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+  BN.LID x4++, 320(x0)
+  BN.LID x4++, 352(x0)
+
+  JAL x1, barret384
```
typo
```suggestion
  JAL x1, barrett384
```

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+ *                               here).
+ * @param[out] [559:512]: c, result, max. length 384 bit.
+ */
+wrap_barret384:
```
typo
```suggestion
wrap_barrett384:
```

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+ * Clobbered registers: w16, w17, w18, w19, w20, w21, w22, w23, w24
+ * Clobbered flag groups: standard
+ */
+barret384:
```
typo
```suggestion
barrett384:
```

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+  BN.SEL w29, w21, w29, c
+  BN.SEL w30, w22, w30, c
+
+  /* return result: [w29, w30] =  in1*in2 % m. */
```
```suggestion
  /* return result: [w29, w30] =  c = a * b % m. */
```

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+  BN.SUBI w30, w31, 1
+  BN.RSHI w30, w30, w31 >> 127
+
+  /* Compute the integer product of the operands x = in1 * in2
```
```suggestion
  /* Compute the integer product of the operands x = a * b
```

felixmiller commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,280 @@
+  BN.RSHI w30, w30, w31 >> 127
+
+  /* Compute the integer product of the operands x = in1 * in2
+     x = [w18, w17, w16] = in1 * in2 = [w9, w8] * [w11, w10]
```
```suggestion
     x = [w18, w17, w16] = a * b = [w9, w8] * [w11, w10]
```

rswarbrick commented:
This is really neat! I'm not claiming I followed the maths properly, but it looks sensible to me. It also makes me wish we had a register allocator (I wrote one in another life at Broadcom, which would be lovely to drop in here!).

I think the commit currently at the top of the PR branch (8d78d9b) still has syntax that chokes the assembler. Could you fix that? If you rebase onto master, make sure you also cherry-pick #3270 if it isn't yet merged: I slightly broke the assembler and this fixes it!

gkelly commented:
This is a very cool implementation, and I feel a bit bad for being That Guy, but we do have an assembly style guide: https://docs.opentitan.org/doc/rm/asm_coding_style/. I think it's probably worth following it in OTBN given that it's RISC-V-enough that that probably covers it. Uniformity across the project is a very useful property.

felixmiller commented:
@gkelly No worries, absolutely happy to receive some guidance here. Actually I tried to follow the style guide you linked, can you elaborate a bit?

felixmiller commented:
@rswarbrick Could you have another quick look?
Changes:
- Fixed all the flag parameters to uppercase now (I think some slipped on the first pass)
- Replaced the multiplication kernel by one based on Greg's fully unrolled version
- Added the section directive
- Added ecall intstruction
- Fixed some typos

rswarbrick commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,266 @@
+ * Clobbered flag groups: standard
+ */
+mul384:
+  BN.MULQACC.Z         w8.0, w10.0,   0
```
Just a note: I really like the formatting in this block!

rswarbrick reviewed (approved):
I guess we need to wait for Garret to check his thoughts about style, but this looks good to me.

imphil commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,266 @@
+ * @param[out] [w18, w17, w16]: c, result, max. length 768 bit.
+ *
+ * Clobbered registers: w19, w20
+ * Clobbered flag groups: standard
```
What do you mean by "standard" here? Is it flag group 0?

felixmiller replied:
Yes, I mean't FG0. You're right, we stopped using the terms "standard" and "extension" for this. Updated the PR

mcy commented:
I'm not sure what Garret's thoughts on style are, but the main thing I would take a look at is consistency with existing assembly files (https://github.com/lowRISC/opentitan/blob/master/sw/device/exts/common/flash_crt.S is a good example).

The following are de-facto style but we probably aught to include them in the style guide:
- Lower-case mnemonics and register names (so as not to clash with macros; same reason for `kConst` in C/C++).
- `//` for non-Doxygen comments, including the license header.

In addition, the style guide requires the use of pseudo instructions when possible, Given that the OBTN assembler is implemented on top of clang, I think this guidance applies here (e.g., prefer `ret` over `jalr zero, ra, 0`. This also applies to sequences like `addi rd, zero, 8`; prefer `li 8`, unless it is known that `li` produces more code than is appropriate (lui + addi instead of just an addi).
There is also a provision for using ABI names for RISC-V GPAs (`ra` over `x1`) that I feel applies here (though I can't remember if `x0` is hard-wired to zero in OBTN or not, so maybe it's ok to just call it `x0`?)

Of course, since OBTN isn't exactly RISC-V, I feel it is completely valid to include provisions in the style guide specifically for it.

lenary commented:
@mcy: In reply to your second bullet, `otbn-as` is not being run via gcc/clang (which strips comments), so it actually only supports `/* */` style comments. (Meson runs `.S` files through the C compiler).

To clarify, `otbn-as` is a small wrapper on top of binutils' `as`. Lots of pseudo-instructions are not implemented in the assembler yet, and its not clear there will be support for all the same pseudo-instructions anyway, as I understand it.

rswarbrick commented:
Sam's right. At the moment, the only pseudo-instructions we support in `otbn-as` are `li`, `ret` and `nop`. Glancing through the code, I can spot two candidates for `li` (lines 245, 261) and two for `ret` (lines 74, 222), but I'm not sure how much this matters.

mcy commented:
Ok, that makes sense! I'm ok submitting this PR as-is, but we should probably update the style guide with a "OTBN exceptions/weirdnesses" section, since I didn't know any of those details, which is why I didn't style-review this PR in the first place.

felixmiller commented:
- **Pseudo instructions:** So far I didn't use them, because it is still a bit of a moving target and we will have to update the code anyways at some point. So I thought let's have one change later updating everything to use the PIs. Anyways, changed the current PR to use `ret` and `li` now.

- **Mnemonic letter case:** Also changed this but am having 2nd thoughts now and didn't update the PR yet. Wanted to get your view first. See [here](https://github.com/felixmiller/opentitan/blob/otbn-barrett-lc/sw/otbn/code-snippets/barrett384.S) for a lower case variant. I am finding the UC variant much more readable. You can spot the actual instructions much better between the comments and most crypto code will look like this and contain formulas in the comments that look a bit like instructions on the first glance. I am absolutely seeing @mcy's point w.r.t. macros though. Another option would be to go for doxygen style comments everywhere (when it's multi line) to have each comment line start with a `*`.

- **Style guide:** Absolutely agree that we should come up with one for OTBN at some point. From my point of view it should be a self-contained one though with as much consistency as possible with the RISC-V one. Then have a link at the top of the RISC-V one: "If you are actually writing OTBN code go there". We are not there yet there though and just having the first discussions on style.

rswarbrick commented:
@felixmiller: Good idea about the cross links for a style guide.

Re upper and lower case mnemonics, I probably prefer the lower case because it looks less "shouty", but I don't have strong feelings. I think both options are pretty painful to read without syntax highlighting: I'm not convinced that uppercase mnemonics help much.

mcy commented:
Re: upper case, I mostly think we should stick to it for consistency. I myself have trouble reading instructions that are upper-case, so I think we aught to just pick whatever is consistent (FWIW, a lot of our tooling prefers lower-case: both objdump and the Ibex instruction trace are in lower-case). We don't absolutely need to make it lowercase in this PR, but I agree with @rswarbrick that it's going to be fairly hard to read without syntax highlighting (and, in my opinion, `//` comments) anyway.

felixmiller commented:
Thanks @rswarbrick and @mcy for your feedback. Changed mnemonics to lower case now.

imphil commented:
CI fails with
`../../../../sw/otbn/code-snippets/barrett384.S:35: Invalid option value, '.z'. If specified, it should have been '.Z'.`

@rswarbrick should we allow case-insensitive/lower-case options like `.z`? I think yes.

felixmiller commented:
Ah, I am just seeing that the `.z` is upper case in the other snippets as well:
https://github.com/lowRISC/opentitan/blob/master/sw/otbn/code-snippets/mul384.S#L30

Is there a reason why `.so` is lower and `Z` is upper? Thought both was part of the mnemonic

gkelly commented:
Thanks for following up here @mcy. With the mnemonics updated and the correct pseudoinstructions I think this looks good to me.

rswarbrick commented:
Ah, the difference is that we have 3 mnemonics for bignum multiplication: `mulqacc`, `mulqacc.so` and `mulqacc.wo`. The `.z` bit is technically an "option operand": something that is encoded as 1 if present and 0 if not. (There's some magic called "glued-ops" that allows you to write `mulqacc.z` rather than `mulqacc .z`)

As it stands, `obtn-as` does case-folding when checking mnemonics, but not when checking option operands or enums. This seems a bit silly to me: I'll push up a PR in a minute to case-fold those too.

rswarbrick commented:
That turned out to be easy! Case-folding PR at #3324.

imphil commented:
This looks good to go, let's get it in!

cdgori commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,265 @@
+  bn.add w30, w30, w22
+  bn.sel w22, w30, w22, C
+
+  /* Barret algorithm requires subtraction of the modulus at most two times if
```
Sorry for very late (post-commit) nit, but: Barret -> Barrett

cdgori commented on sw/otbn/code-snippets/barrett384.S:
```diff
@@ -0,0 +1,265 @@
+  /* load modulus from dmem to [w13, w12] */
+  bn.lid x4++, 256(x0)
+  bn.lid x4++, 288(x0)
+  /* load barret precomputed parameter u from dmem to [w15, w14] */
```
also barret -> barrett
